toolchain/check/eval.cpp

```diff
@@ -1846,6 +1880,35 @@ auto TryEvalTypedInst<SemIR::WhereExpr>(EvalContext& eval_context,
         UpdatePhaseIgnorePeriodSelf(eval_context, rhs, &phase);
         info.rewrite_constraints.push_back(
             {.lhs_const_id = lhs, .rhs_const_id = rhs});
+      } else if (auto impls =
+                     eval_context.insts().TryGetAs<SemIR::RequirementImpls>(
+                         inst_id)) {
+        SemIR::ConstantId lhs = eval_context.GetConstantValue(impls->lhs_id);
+        SemIR::ConstantId rhs = eval_context.GetConstantValue(impls->rhs_id);
+        if (rhs != SemIR::ErrorInst::SingletonConstantId &&
+            IsPeriodSelf(eval_context, lhs)) {
+          auto rhs_inst_id = eval_context.constant_values().GetInstId(rhs);
+          if (eval_context.insts().Is<SemIR::TypeType>(rhs_inst_id)) {
+            // `.Self impls type` -> nothing to do.
+          } else {
+            auto facet_type =
+                eval_context.insts().GetAs<SemIR::FacetType>(rhs_inst_id);
+            SemIR::FacetTypeInfo more_info = GetConstantFacetTypeInfo(
+                eval_context, facet_type.facet_type_id, &phase);
+            // All interface requirements go in `self_impls_constraints`.
+            llvm::append_range(info.self_impls_constraints,
+                               more_info.extend_constraints);
+            llvm::append_range(info.self_impls_constraints,
+                               more_info.self_impls_constraints);
+            // Other requirements are copied in.
```

Comment:
Do these need to be associated with the interface where they were rewriting in some way?

`impl C as Z where .Self impls (Y where .A = i32);`

Here we end up with a FacetTypeInfo with a rewrite constraint of simply `.A = i32`, right, which would be like a rewrite of `Z.A` instead of `Y.A`?

`impl C as Z where .A = u32 and .Self impls (Y where .A = i32);`

And here we'd have two rewrites of `Z.A` to different values and none of `Y.A`?

Is there something we can do now, or a TODO and some tests we can write that fail? Or does this somehow work already and we can show that in a test?